### PR TITLE
feat: Support `COURSIER_SHA256` environment variable

### LINKED ISSUE #1526: [feat] Support `COURSIER_SHA256` environment variable
## Summary

Add `COURSIER_SHA256` environment variable to allow users to specify a different Coursier version, complementing the existing `COURSIER_URL` env var.

## Current State

[PR #402](https://github.com/bazel-contrib/rules_jvm_external/pull/402) added `COURSIER_URL` to allow downloading Coursier from a custom location. However, the SHA256 checksum is hardcoded, which means:

- ✅ You **can** change **where** to download Coursier (mirror, internal repo, local file)
- ❌ You **cannot** change **which version** of Coursier to use

From [`private/rules/coursier.bzl:1163`](https://github.com/bazel-contrib/rules_jvm_external/blob/master/private/rules/coursier.bzl#L1163):
```python
repository_ctx.download(coursier_download_urls, "coursier", sha256 = COURSIER_CLI_SHA256, executable = True)
```

In the original discussion ([#389](https://github.com/bazel-contrib/rules_jvm_external/issues/389#issuecomment-611689342)), @jin suggested that both `COURSIER_URL` and `COURSIER_SHA256` would follow the existing pattern:

> Since we're already using an environment variable for `COURSIER_CACHE`, `COURSIER_URL` and `COURSIER_SHA256` would work to replace the values here

However, only `COURSIER_URL` was implemented in [PR #402](https://github.com/bazel-contrib/rules_jvm_external/pull/402).

### 1. Testing Coursier milestones

Since v2.1.24 (current in rules_jvm_external), Coursier has published **23 milestones** with fixes that users may need:

| Milestone | Notable Changes |
|-----------|-----------------|
| v2.1.25-M22 | [#3534](https://github.com/coursier/coursier/pull/3534) - Sensible defaults for Gradle modules |
| v2.1.25-M22 | [#3539](https://github.com/coursier/coursier/pull/3539) - Accept rejects in GMM version specifiers |
| v2.1.25-M16 | [#3402](https://github.com/coursier/coursier/pull/3402) - GMM endorseStrictVersions support |
| v2.1.25-M15 | [#3387](https://github.com/coursier/coursier/pull/3387) - Source JAR support for Gradle Modules |
| v2.1.25-M14 | [#3384](https://github.com/coursier/coursier/pull/3384) - Various Gradle Modules fixes |

Users working with Kotlin Multiplatform, Jetpack Compose, or AndroidX dependencies often hit GMM edge cases fixed in milestones.

### 2. Testing patches before contributing upstream

When users encounter Coursier bugs, being able to easily test a patched version:
- Validates the fix works in their environment
- Makes it easier to contribute patches back to Coursier
- Reduces the feedback loop between bug report and verification

### 3. Using custom Coursier builds

- Pre-release versions for early testing
- Organization-specific patches
- Pinning to a specific commit for reproducibility

### 4. Air-gapped environments

Internal artifact repositories may host approved Coursier versions that differ from the one hardcoded in rules_jvm_external.

## Proposed Change

```python
# In environ list:
environ = [
    # ... existing ...
    "COURSIER_URL",
    "COURSIER_SHA256",  # NEW
]

# In _coursier_fetch_impl:
coursier_sha256 = repository_ctx.os.environ.get("COURSIER_SHA256", COURSIER_CLI_SHA256)
repository_ctx.download(coursier_download_urls, "coursier", sha256 = coursier_sha256, executable = True)
```

## Usage

```bash
# Use a Coursier milestone
bazel build //... \
  --repo_env=COURSIER_URL="https://github.com/coursier/coursier/releases/download/v2.1.25-M23/coursier.jar" \
  --repo_env=COURSIER_SHA256="<sha256-of-M23>"

# Use a local build
bazel build //... \
  --repo_env=COURSIER_URL="file:///path/to/custom/coursier.jar" \
  --repo_env=COURSIER_SHA256="$(shasum -a 256 /path/to/custom/coursier.jar | cut -d' ' -f1)"
```

## Current Workaround

Users must patch `private/versions.bzl` via `single_version_override`:

```python
# MODULE.bazel
single_version_override(
    module_name = "rules_jvm_external",
    patches = ["//patches:rules_jvm_external_custom_coursier.patch"],
)
```

```diff
# patches/rules_jvm_external_custom_coursier.patch
--- a/private/versions.bzl
+++ b/private/versions.bzl
@@ -1,5 +1,5 @@
-_COURSIER_CLI_VERSION = "v2.1.24"
+_COURSIER_CLI_VERSION = "v2.1.25-M23"

-COURSIER_CLI_SHA256 = "8c724dc204534353ea8263ba0af624979658f7ab62395f35b04f03ce5714f330"
+COURSIER_CLI_SHA256 = "<new-sha256>"
```

## Alternatives Considered

| Alternative | Notes |
|-------------|-------|
| Patch `versions.bzl` (current workaround) | Works, but requires a patch file |
| `--override_repository` | Requires maintaining a full fork of rules_jvm_external |
| Toolchain-based approach | [Previously discussed](https://github.com/bazel-contrib/rules_jvm_external/issues/389#issuecomment-612190570) - "coursier is an implementation detail" |
| `COURSIER_SHA256` env var | Proposed here - follows existing `COURSIER_URL` pattern |

## Notes

- Users who override the Coursier version accept that rules_jvm_external was not tested with that version
- SHA256 verification is preserved - users must provide the correct checksum
- Follows the pattern established by `COURSIER_URL`
- ~5 line change, backwards compatible

### PR DESCRIPTION
**Pull request by albertocavalcante**
Allows users to specify a different Coursier version by overriding the SHA256 checksum, complementing the existing `COURSIER_URL` env var.

## Summary

- Add `get_coursier_sha256()` helper function (testable)
- Use env var override in `_coursier_fetch_impl`
- Add `COURSIER_SHA256` to environ list
- Add 2 unit tests
- Update README with usage and compatibility note

## Changes

```python
# New helper function
def get_coursier_sha256(environ, default_sha256):
    return environ.get("COURSIER_SHA256", default_sha256)

# Usage in _coursier_fetch_impl
coursier_sha256 = get_coursier_sha256(repository_ctx.os.environ, COURSIER_CLI_SHA256)
repository_ctx.download(coursier_download_urls, "coursier", sha256 = coursier_sha256, executable = True)
```

## Usage

```bash
bazel build @maven//... \
  --repo_env=COURSIER_URL='https://github.com/coursier/coursier/releases/download/v2.1.25-M23/coursier.jar' \
  --repo_env=COURSIER_SHA256='<sha256-of-the-jar>'
```

Closes #1526

### CODE REVIEW

**Comment by albertocavalcante** [comment]
Apologies for the follow-up. I understand you must be busy, but PTAL whenever you get a chance. 🙏🏻

@shs96c @jin @cheister 

It would be great to get feedback and know whether you're OK with this change or if you'd advise taking another approach. Thanks!
